src/ripple/app/misc/AMM_formulae.h

```diff
@@ -0,0 +1,314 @@
+ * @param lpTokensBalance LP Tokens balance
+ * @param tfee trading fee in basis points
+ * @return tokens out amount
+ */
+STAmount
+lpTokensOut(
+    STAmount const& asset1Balance,
+    STAmount const& asset1Withdraw,
+    STAmount const& lpTokensBalance,
+    std::uint16_t tfee);
+
+/** Calculate asset withdrawal by tokens
+ * @param assetBalance balance of the asset being withdrawn
+ * @param lptAMMBalance total AMM Tokens balance
+ * @param lpTokens LP Tokens balance
+ * @param tfee trading fee in basis points
+ * @return calculated asset amount
+ */
+STAmount
+withdrawByTokens(
+    STAmount const& assetBalance,
+    STAmount const& lptAMMBalance,
+    STAmount const& lpTokens,
+    std::uint32_t tfee);
+
+/** Find in/out amounts to change the spot price quality to the requested
+ * quality.
+ * @param pool AMM pool balances
+ * @param quality requested quality
+ * @param tfee trading fee in basis points
+ * @return seated in/out amounts if the quality can be changed
+ */
+template <typename TIn, typename TOut>
+std::optional<TAmounts<TIn, TOut>>
+changeSpotPriceQuality(
+    TAmounts<TIn, TOut> const& pool,
+    Quality const& quality,
+    std::uint32_t tfee)
+{
+    if (auto const nTakerPays =
+            (root2(pool.in * pool.out * quality.rate() / feeMult(tfee)) -
+             pool.in / feeMult(tfee));
+        nTakerPays > 0)
+    {
+        auto const takerPays = toAmount<TIn>(
+            getIssue(pool.in), nTakerPays, Number::rounding_mode::upward);
+        // should not fail
+        if (auto const amounts =
+                TAmounts<TIn, TOut>{
+                    takerPays, swapAssetIn(pool, takerPays, tfee)};
+            Quality{amounts} < quality)
+            Throw<std::runtime_error>("changeSpotPriceQuality failed");
+        else
+            return amounts;
+    }
+    return std::nullopt;
+}
+
+/** AMM pool invariant - the product (A * B) after swap in/out has to remain
+ * at least the same: (A + in) * (B - out) >= A * B
+ * XRP round-off may result in a smaller product after swap in/out.
+ * To address this:
+ *   - if on swapIn the out is XRP then the amount is round-off
+ *     downward, making the product slightly larger since out
+ *     value is reduced.
+ *   - if on swapOut the in is XRP then the amount is round-off
+ *     upward, making the product slightly larger since in
+ *     value is increased.
+ */
+
+/** Swap assetIn into the pool and swap out a proportional amount
+ * of the other asset.
+ * @param pool current AMM pool balances
+ * @param assetIn amount to swap in
+ * @param tfee trading fee in basis points
+ * @return
+ */
+template <typename TIn, typename TOut>
+TOut
+swapAssetIn(
+    TAmounts<TIn, TOut> const& pool,
+    TIn const& assetIn,
+    std::uint16_t tfee)
+{
+    return toAmount<TOut>(
+        getIssue(pool.out),
+        pool.out - (pool.in * pool.out) / (pool.in + assetIn * feeMult(tfee)),
+        Number::rounding_mode::downward);
```

Comment:
I think this is equation 9 from the spec. I also think it would be a good idea, since we have a published spec, to add comments to this and other semi-opaque formulas specifying (by number) which equation is being used. Not only will it help with this review, but it will also help future-proof the code so that people unfamiliar with the code and/or spec can jump right to the relevant part of the spec for an explanation.